File: server/models/audit.py
"""Audit trace model — full pipeline traceability."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import String, Text, DateTime, JSON, Float
from sqlalchemy.orm import Mapped, mapped_column

from server.db import Base


class AuditTrace(Base):
    """A single trace record capturing one event in the pipeline."""

    __tablename__ = "audit_traces"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    # Correlation id for end-to-end tracking
    trace_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)
    session_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)
    agent_id: Mapped[str] = mapped_column(String(36), nullable=False)
    tenant_id: Mapped[str] = mapped_column(String(36), default="default")

    # Event classification
    event_type: Mapped[str] = mapped_column(
        String(32), nullable=False
    )  # user_input | intent | retrieval | llm_call | tool_call | workflow_step | response | escalation | error

    # Event data (flexible JSON)
    event_data: Mapped[dict | None] = mapped_column(JSON, nullable=True)

    # For retrieval events: which chunks were hit
    retrieval_hits: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # For LLM events: model, prompt tokens, completion tokens
    llm_meta: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # For tool events: tool_id, input, output, success
    tool_meta: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # For workflow events: step_id, status, collected fields
    workflow_meta: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    # For escalation: reason
    escalation_reason: Mapped[str | None] = mapped_column(Text, nullable=True)

    # Timing
    latency_ms: Mapped[float | None] = mapped_column(Float, nullable=True)
    timestamp: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

File: console/src/types/audit.ts
export interface AuditTrace {
  id: string;
  trace_id: string;
  session_id: string;
  agent_id: string;
  tenant_id: string;
  event_type: string;
  event_data: Record<string, unknown> | null;
  retrieval_hits: Record<string, unknown> | null;
  llm_meta: Record<string, unknown> | null;
  tool_meta: Record<string, unknown> | null;
  workflow_meta: Record<string, unknown> | null;
  escalation_reason?: string;
  latency_ms?: number;
  timestamp?: string;
}

export interface AuditTraceSummary {
  id: string;
  trace_id: string;
  session_id: string;
  agent_id: string;
  event_type: string;
  latency_ms: number | null;
  timestamp: string | null;
}

export interface AuditTraceListResponse {
  total: number;
  offset: number;
  limit: number;
  items: AuditTraceSummary[];
}

export interface AuditMetrics {
  total_requests: number;
  avg_latency_ms: number;
  error_rate: number;
  escalation_rate: number;
  top_agents: Array<{ agent_id: string; count: number }>;
}
